doc/man/man5/slconf.5.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man5/slconf.5.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH SLCONF 5 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ System File Formats"
.SH NAME
.B slconf
\- Signalling Link Configuration
.\"
.\"
.SH SYNOPSIS
.PP
.B /etc/slconf
.PP
.B #include <sys/slconf.h>
.\"
.\"
.SH DESCRIPTION
.PP
The signaling link configuration database file,
.BR /etc/slconf ,
is a system file that contains configuration information for signalling links and, specifically, the
Signalling Link configuration daemon,
.BR slconfd (8).
.PP
Each entry in the signalling link configuration database file can be represented by a
.B slconfig
structure, formatted as follows:
.\"
.\"
.SH "FILE FORMAT"
.PP
Entries consist of a number of fields separated by newlines.  Fields are separated by whitespace and
occur in the order in which they are described below.  Whitespace can be embedded as blanks or
tabs.  Backslashes may be embedded as backslashes.  Lines that begin with an octothorpe,
.RI ` # ',
in column 1 are treated as comments.  Each valid line in the file describes an available signalling
link its associated configuration information.  Each entry contains the following fields:
.PP
\fBsl\fP <\fIclei\fP> <\fIglobal-ppa\fP> <\fIdevice-type\fP> \fI...\fP
.PP
Where,
.TP
.I clei
A Common Language Equipment Identifier (CLEI) for the signalling link.  This CLEI can be used by
streams opening the
.BR sl-mux (4)
driver to attach to this signalling link.
.TP
.I global-ppa
A global Physical Point of Attachment (PPA) for the signalling link.  This PPA can be used by
streams opening the
.BR sl-mux (4)
driver to attach to this signalling link.
.TP
.I device-type
The device type defines the type of device that a is providing the signalling link.  The device type
can, for example, be one of:
.BR x400p ", " m2ua " or " m2pa .
For additional information on X400P-SS7 configuration, see
.BR x400pconf (8).
For additional information on M2PA configuration, see
.BR m2paconfig (8).
For additional information on M2UA configuration, see
.BR m2uaconfig (8).
.TP
.I ...
The remaining fileds are interpreted depending on the device type.  The signalling link
configuration daemon,
.BR slconfd (8),
that reads this file will attempt to open the shared object object derived from the name of the
device type.  The
.BR slconfd (8)
program will attempt to open the shared object named
\(lq\fB/usr/lib/ss7/\fIdevice-type\fB.so\fR\(rq,
where
.I device-type
is the name of the device type from the
.I device-type
field.
For example, for the
.B x400p
device type, the
.BR slconfd (8)
program will attempt to open
.RB \(lq /usr/lib/ss7/x400p.so \(rq.
.IP ""
The shared object opened by
.BR slconfd (8)
provides access procedures for the daemon to open, attach and configure the signalling link.  The
access procedures are passed the remainder of the configuration file entry and is parsed according
to the specific device type.
.PP
\fBss7profile\fR <\fIname\fR> <\fIvariant\fR> <\fIoptions\fR>
.PP
Where,
.TP
.B ss7profile
is an entry selector that indicates that this is an SS7 profile entry.  An SS7 profile is a
collection of protocol variant and option information for SS7 entities that is commont to a number
of SS7 entities.
.TP
.I name
is an identifier for the SS7 profile.
.TP
.I variant
specifies a protocol variant.  A protocol variant identifies a standard to which the SS7 entity must
comply.
.TP
.I options
specifies protocol options.  Protocol options are optional behaviours within a standard.  This is a
comma separated list of the options which apply to the protocol variant.
.PP
\fBsdlprofile\fR <\fIname\fR> \fI...\fR
.PP
Where,
.TP
.B sdlprofile
is an entry selector that indicates that this is a signalling data link profile entry.
.TP
.I name
is an identifier for the signalling data link profile.
.TP
.I iftype
specifies the interface type.  The interface type can be one of the following:
.sp
.RS
.TP
.B -
the interface type is unspecified.  The default interface type for the signalling data link will be
used.
.PD 0
.TP
.B v35
specifies a V.35 interface.
.TP
.B ds0
specifies a DS0 interface.
.TP
.B ds0a
specifies a DS0A interface.
.TP
.B e1
specifies an (full span) E1 interface.
.TP
.B t1
specifies a (full span) T1 interface.
.TP
.B j1
specifies a (full span) J1 interface.
.TP
.B atm
specifies an ATM (SSCOP) interface.
.TP
.B packet
specifies a packet interface.
.PD
.RE
.TP
.I ifrate
specifies the interface rate.  The interface rate can be one of the following:
.sp
.RS
.TP
.B -
the interface rate is unspecified.  The default interface rate for the signalling data link will be
used.
.PD 0
.TP
.B ds0a
specifies a DS0A interface rate (56000 bps).
.TP
.B ds0
specifies a DS0 interface rate (64000 bps).
.TP
.B t1
specifies a T1 interface rate (1544000 bps).
.TP
.B j1
specifies a J1 interface rate (1544000 bps).
.TP
.B e1
specifies an E1 interface rate (2048000 bps).
.PD
.RE
.TP
.I ifgtype
specifies the interface group type.  The interface group type can be one of the following:
.sp
.RS
.RE
.TP
.I ifgrate
specifies the interface group rate.  The interface group rate can be one of the following:
.sp
.RS
.RE
.TP
.I ifmode
specifies the interface mode.  The interface mode can be one of the following:
.sp
.RS
.RE
.TP
.I ifgmode
specifies the interface group mode.  The interface group mode can be one of the following:
.sp
.RS
.RE
.TP
.I ifgcrc
specifies the interface group CRC.  The interface group CRC can be one of the following:
.sp
.RS
.RE
.TP
.I ifclock
specifies the interface clock.  The interface clock can be one of the following:
.sp
.RS
.RE
.TP
.I ifcoding
specifies the interface coding.  The interface coding can be one of the following:
.sp
.RS
.RE
.TP
.I ifframing
specifies the interface framing.  The interface framing can be one of the following:
.sp
.RS
.RE
.TP
.I ifblksize
specifies the interface block size.  The interface block size can be one of the following:
.sp
.RS
.RE
.TP
.I ifrxlevel
specifies the interface receiver level.  The interface receiver level can be one of the following:
.sp
.RS
.RE
.TP
.I iftxlevel
specifies the interface transmitter level.  The interface transmitter level can be one of the following:
.sp
.RS
.RE
.TP
.I ifsyncsrc
specifies the interface synchronization sources.  The interface synchronization sources is a comma
separated list, with no intervening whitespace, specifying the synchronization sources in order of
priority for the interface.  Each interface synchronization source can be one of the following:
.sp
.RS
.RE
.PP
\fBsdtprofile\fR <\fIname\fR> \fI...\fR
.PP
Where,
.TP
.B sdtprofile
is an entry selector that indicates that this is a signalling data terminal profile entry.
.TP
.I name
is an identifier for the signalling data terminal profile.
.PP
\fBslprofile\fR <\fIname\fR> \fI...\fR
.PP
Where,
.TP
.B slprofile
is an entry selector that indicates that this is a signalling link profile entry.  A signalling
link profile is a collection of configuration information for signalling links that are common to a
number of signalling links.
.TP
.I name
is an identifier for the signalling link profile.
.\"
.\"
.SH FILES
.PP
.TP
.B @configdir@/slconf
The signalling link configuration file.
.\"
.\"
.SH BUGS
.PP
Not done yet.
.\"
.[
$LIST$
.]
.\"
.TA


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/util/slconf.c
/*****************************************************************************

 @(#) File: src/util/slconf.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/util/slconf.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

#include <stropts.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <stddef.h>
#include <stdint.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/poll.h>
#include <sys/time.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <getopt.h>
#include <time.h>
#include <signal.h>
#include <syslog.h>
#include <sys/utsname.h>

static int verbose = 1;
static int debug = 0;

#ifndef MAX_PATH_LENGTH
#define MAX_PATH_LENGTH 256
#endif

#ifndef SLCONF_CONF_FILENAME
#define SLCONF_CONF_FILENAME "/etc/sysconfig/slconf.conf"
#endif

char filename[MAX_PATH_LENGTH] = SLCONF_CONF_FILENAME;

enum {
	SOURCE_NONE = 0,
	SOURCE_STDIN,
	SOURCE_FILE,
	SOURCE_CMDLINE
} source = SOURCE_NONE;

int interactive = 0;

void
copying(int argc, char *argv[])
{
	if (!verbose)
		return;
	(void) fprintf(stdout, "\
--------------------------------------------------------------------------------\n\
%1$s\n\
--------------------------------------------------------------------------------\n\
Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>\n\
Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>\n\
Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>\n\
\n\
All Rights Reserved.\n\
--------------------------------------------------------------------------------\n\
This program is free software: you can  redistribute it  and/or modify  it under\n\
the terms of the  GNU Affero  General  Public  License  as published by the Free\n\
Software Foundation, version 3 of the license.\n\
\n\
This program is distributed in the hope that it will  be useful, but WITHOUT ANY\n\
WARRANTY; without even  the implied warranty of MERCHANTABILITY or FITNESS FOR A\n\
PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.\n\
\n\
You should have received a copy of the  GNU Affero General Public License  along\n\
with this program.   If not, see <http://www.gnu.org/licenses/>, or write to the\n\
Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n\
--------------------------------------------------------------------------------\n\
U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on behalf\n\
of the U.S. Government (\"Government\"), the following provisions apply to you. If\n\
the Software is supplied by the Department of Defense (\"DoD\"), it is classified\n\
as \"Commercial  Computer  Software\"  under  paragraph  252.227-7014  of the  DoD\n\
Supplement  to the  Federal Acquisition Regulations  (\"DFARS\") (or any successor\n\
regulations) and the  Government  is acquiring  only the  license rights granted\n\
herein (the license rights customarily provided to non-Government users). If the\n\
Software is supplied to any unit or agency of the Government  other than DoD, it\n\
is  classified as  \"Restricted Computer Software\" and the Government's rights in\n\
the Software  are defined  in  paragraph 52.227-19  of the  Federal  Acquisition\n\
Regulations (\"FAR\")  (or any successor regulations) or, in the cases of NASA, in\n\
paragraph  18.52.227-86 of  the  NASA  Supplement  to the FAR (or any  successor\n\
regulations).\n\
--------------------------------------------------------------------------------\n\
Commercial  licensing  and  support of this  software is  available from OpenSS7\n\
Corporation at a fee.  See http://www.openss7.com/\n\
--------------------------------------------------------------------------------\n\
", ident);
}

void
version(int argc, char *argv[])
{
	if (!verbose)
		return;
	(void) fprintf(stdout, "\
%1$s (OpenSS7 %2$s) %3$s (%4$s)\n\
Written by Brian Bidulock.\n\
\n\
Copyright (c) 2008, 2009, 2010, 2011, 2015  Monavacon Limited.\n\
Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008  OpenSS7 Corporation.\n\
Copyright (c) 1997, 1998, 1999, 2000, 2001  Brian F. G. Bidulock.\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\
\n\
Distributed by OpenSS7 under GNU Affero General Public License Version 3,\n\
with conditions, incorporated herein by reference.\n\
\n\
See `%1$s --copying' for copying permissions.\n\
", NAME, PACKAGE, VERSION, PACKAGE_ENVR " " PACKAGE_DATE);
}

void
usage(int argc, char *argv[])
{
	if (!verbose)
		return;
	(void) fprintf(stderr, "\
Usage:\n\
    %1$s [options] {-i|--interactive}\n\
    %1$s [options] {-f|--file}[=FILENAME]\n\
    %1$s [options] COMMAND\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
", argv[0]);
}

void
help(int argc, char *argv[])
{
	if (!verbose)
		return;
	(void) fprintf(stdout, "\
Usage:\n\
    %1$s [options] [{-c|--command}] -- COMMAND[; COMMAND]*\n\
    %1$s [options] {-f|--file}[=FILENAME]\n\
    %1$s [options] {-i|--interactive}\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
Arguments:\n\
    COMMAND\n\
General Options:\n\
    -q, --quiet\n\
        suppress normal output (equivalent to --verbose=0)\n\
    -d, --debug [LEVEL]\n\
        increment or set debug LEVEL [default: 0]\n\
    -v, --verbose [LEVEL]\n\
        increment or set output verbosity LEVEL [default: 1]\n\
        this option may be repeated.\n\
Command Options:\n\
    -c, --command\n\
        process remainder of arguments as command\n\
    -f, --file[=FILENAME]               (default: %2$s)\n\
        read configuration commands from file\n\
    -i, --interactive\n\
        enter interactive command shell\n\
    -h, --help, -?, --?\n\
        print this usage information and exit\n\
    -V, --version\n\
        print version and exit\n\
    -C, --copying\n\
        print copying permission and exit\n\
", argv[0], filename);
}

extern void yyinit(char *name, FILE *input, FILE *output);
extern int yyparse(void);

void
slconf_stdin(char *name)
{
    yyinit(name, stdin, stdout);
    while (yyparse() == 0) ;
    return;
}

void
slconf_file(char *name, const char *filename)
{
	FILE *f;

	if ((f = fopen(filename, "r")) == NULL) {
		perror(name);
		exit(1);
	}
	yyinit(name, f, stdout);
	while (yyparse() == 0) ;
	fclose(f);
	return;
}

int
main(int argc, char *argv[])
{
	int c;
	int val;
	char *optstr;

	while (1) {
		int option_index = 0;
		/* *INDENT-OFF* */
		static struct option long_options[] = {
			{"command",	no_argument,		NULL,	'c'},
			{"file",	optional_argument,	NULL,	'f'},
			{"interactive",	no_argument,		NULL,	'i'},
			{"debug",	optional_argument,	NULL,	'd'},
			{"verbose",	optional_argument,	NULL,	'v'},
			{"quiet",	no_argument,		NULL,	'q'},
			{"help",	no_argument,		NULL,	'h'},
			{"version",	no_argument,		NULL,	'V'},
			{"copying",	no_argument,		NULL,	'C'},
			{"?",		no_argument,		NULL,	'h'},
			{NULL,		0,			NULL,	 0 }
		};
		/* *INDENT-ON* */

		c = getopt_long(argc, argv, ":W:cf::id::v::qhVC?", long_options, &option_index);
		if (c == -1)
			break;
		switch (c) {
		case 'c':
			if (source != SOURCE_NONE)
				goto bad_option;
			source = SOURCE_CMDLINE;
			break;
		case 'f':	/* -f, --file [FILENAME] */
			if (source != SOURCE_NONE)
				goto bad_option;
			if (optarg != NULL)
				strncpy(filename, optarg, sizeof(filename));
			source = SOURCE_FILE;
			break;
		case 'i':	/* -i, --interactive */
			if (source != SOURCE_NONE)
				goto bad_option;
			source = SOURCE_STDIN;
			break;
		case 'd':	/* -d, --debug [LEVEL] */
			if ((optstr = optarg) == NULL) {
				debug++;
				break;
			}
			val = strtol(optarg, &optstr, 0);
			if (optstr == optarg || *optstr != '\0' || val < 0)
				goto bad_option;
			debug = val;
			break;
		case 'v':	/* -v, --verbose [LEVEL] */
			if ((optstr = optarg) == NULL) {
				verbose++;
				break;
			}
			val = strtol(optarg, &optstr, 0);
			if (optstr == optarg || *optstr != '\0' || val < 0)
				goto bad_option;
			verbose = val;
			break;
		case 'q':	/* -q, --quiet */
			verbose -= verbose > 1 ? 1 : verbose;
			break;
		case 'h':	/* -h, --help [WHAT] */
			if (source != SOURCE_NONE)
				goto bad_option;
			help(argc, argv);
			exit(0);
		case 'V':	/* -V, --version */
			if (source != SOURCE_NONE)
				goto bad_option;
			version(argc, argv);
			exit(0);
		case 'C':	/* -C, --copying */
			if (source != SOURCE_NONE)
				goto bad_option;
			copying(argc, argv);
			exit(0);
		case ':':	/* missing mandatory parameter */
			optind--;
			fprintf(stderr, "%s: missing parm -- %s\n", argv[0], argv[optind]);
			usage(argc, argv);
			exit(2);
		case '?':	/* illegal option */
		default:
		      bad_option:
			optind--;
		      syntax_error:
			if (optind < argc) {
				fprintf(stderr, "%s: illegal syntax --", argv[0]);
				for (; optind < argc; optind++)
					fprintf(stderr, " %s", argv[optind]);
				fprintf(stderr, "\n");
			}
			usage(argc, argv);
			exit(2);
		}
		if (source == SOURCE_CMDLINE)
			break;
	}
      reevaluate:
	switch (source) {
	case SOURCE_NONE:
		if (optind < argc) {
			source = SOURCE_CMDLINE;
			goto reevaluate;
		}
		source = SOURCE_STDIN;
		goto reevaluate;
#if 0
	case SOURCE_CMDLINE:
		slconf_cmdline(argv[0], argc, argv, optind);
		break;
#endif
	case SOURCE_STDIN:
		if (optind < argc)
			goto syntax_error;
		interactive = 1;
		slconf_stdin(argv[0]);
		break;
	case SOURCE_FILE:
		if (optind < argc)
			goto syntax_error;
		slconf_file(argv[0], filename);
		break;
	default:
		exit(4);
		break;
	}
	exit(0);
}
